Mirroring Kafka - Event Hub blog post

ppatierno opened the pull request:
Signed-off-by: Paolo Patierno <ppatierno@live.com>

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,219 @@
+  name: eventhubssecret
+type: Opaque
+stringData:
+  eventhubsuser: $ConnectionString
```
I think we are not using this. Should we remvoe it?

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,219 @@
+        secretName: eventhubssecret
+        password: eventhubspassword
+    tls:
+      trustedCertificates: []
```
I think this deserves some little explanation in the text - why is it used and how it works.

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,219 @@
+
+### Conclusion
+
+Integrating applications running on different platforms and different clouds is becoming quite common nowadays and the Hybrid Cloud based use cases are growing fast.
```
The Hybrid Cloud here is not a specific product. So I think it should be in lower case

```suggestion
Integrating applications running on different platforms and different clouds is becoming quite common nowadays and the hybrid cloud based use cases are growing fast.
```

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,219 @@
+author: paolo_patierno
+---
+
+Let's imagine to have a software architecture based on Apache Kafka running on an on-premise Kubernetes or OpenShift installation and managed by the Strimzi operators.
```
I wonder if we really need to mention OpenShift here when the whole demo is based on Kubernetes.

ppatierno replied:
Well, I am saying "or OpenShift" as another Kubernetes distribution. During the demo, I had to choose one of course.

scholzj replied:
Which is a bit weird since it doesn't mention the 100 other Kubernetes distributions but just one specific although it doesn't have anything really related to the blogpost..

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,219 @@
+---
+
+Let's imagine to have a software architecture based on Apache Kafka running on an on-premise Kubernetes or OpenShift installation and managed by the Strimzi operators.
+All the services are able to communicate each other exchanging messages through a set of topics.
```
```suggestion
All the services are able to communicate with each other exchanging messages through a set of topics.
```

ppatierno commented:
Agreed. I added a few lines at the beginning of the "Configure Kafka Mirror Maker" paragraph.

scholzj commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -93,6 +93,11 @@ Where `<eventhubs-namespace>` has to be replaced with the actual Event Hub names
 
+When mirroring messages from a source Apache Kafka cluster to a target one, the usual paradigm is deploying in the Kafka Mirror Maker instance alongside the target cluster.
+It means consuming remotely and producing locally.
+In this integration scenario, it is not possible because there is no way to deploy Kafka Mirror Maker alongside the Event Hub namespace that is the target cluster.
```
Would it be worth mentioning that if the reader really wants she/he can run Mirror Maker with Strimzi on AKS to run it in the same cloud? I wouldn't change the blog post because of it, but maybe worth mentioning as alternative? Just thinking ...

ppatierno replied:
I added a sentence about it.

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+author: paolo_patierno
+---
+
+Let's imagine to have a software architecture based on Apache Kafka running on an on-premise Kubernetes installation and managed by the Strimzi operators.
```
```suggestion
Let's imagine having a software architecture based on Apache Kafka running in an on-premise Kubernetes installation and managed by the Strimzi operators.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+All the services are able to communicate with each other exchanging messages through a set of topics.
+At some point there is a new requirement to fullfil: all messages, sent to a specific topic, have to trigger some processing in the cloud via Microsoft Azure Functions already connected to an Azure Event Hub instance.
+It is a third party application already in place that we need to integrate with our own system.
+In order to avoid the producer applications to connect both the on-premise Apache Kafka cluster and to Azure Event Hub, which provides a Kafka protocol head, and sending each message twice, the best solution could be just mirroring the topic.
```
```suggestion
In order to avoid the producing applications having to connect to both the on-premise Apache Kafka cluster and to Azure Event Hub, which provides a Kafka protocol head, and sending each message twice, the best solution could be just mirroring the topic.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+At some point there is a new requirement to fullfil: all messages, sent to a specific topic, have to trigger some processing in the cloud via Microsoft Azure Functions already connected to an Azure Event Hub instance.
+It is a third party application already in place that we need to integrate with our own system.
+In order to avoid the producer applications to connect both the on-premise Apache Kafka cluster and to Azure Event Hub, which provides a Kafka protocol head, and sending each message twice, the best solution could be just mirroring the topic.
+This blog post is going to show how to use the Strimzi cluster operator for configuring and deploying Kafka Mirror Maker in order to mirror topics between two different systems like Apache Kafka and Azure Event Hub.
```
```suggestion
This blog post is going to show how to use the Strimzi cluster operator to configure and deploy Kafka Mirror Maker in order to mirror topics between two different systems like Apache Kafka and Azure Event Hub.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+An application, running on the same cluster, produces messages to an Apache Kafka topic and all the data are consumed by Kafka Mirror Maker in order to be mirrored to the remote Azure Event Hub instance.
+
+On the other side there is the Azure Event Hub namespace, on the Microsoft Azure Cloud, where an Event Hub is filled by the messages mirrored from the corresponding Apache Kafka topic.
+This messages trigger the Azure Functions platform to do the actual processing.
```
```suggestion
These messages trigger the Azure Functions platform to do the actual processing.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+On the other side there is the Azure Event Hub namespace, on the Microsoft Azure Cloud, where an Event Hub is filled by the messages mirrored from the corresponding Apache Kafka topic.
+This messages trigger the Azure Functions platform to do the actual processing.
+For people who don't know about Azure Event Hub, it could be a little bit confusing and not really clear what an Event Hub "namespace" or just an Event Hub mean and how to compare them to an Apache Kafka cluster and topic; it will be clarified in the Event Hub related paragraph of this blog post.
```
You're saying "You might now know what X means, but tough, I'm not going to explain it here. You'll need to be interested enough to read about that later". That's really unhelpful to those readers who don't know about X already. What are your reasons for doing it? I think it would be better to have a paragraph or section where you explain it, and encourage those readers who already know to skip to the next part.

ppatierno replied:
Maybe I can move the section where I explain it from the EH paragraph to here?

sjwoodman replied:
TBH I think you could just remove this sentence.

ppatierno replied:
Yeah, that's a good idea. They will learn more in the EH paragraph if they don't know what it is and how it works.

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+### Strimzi operators: Apache Kafka and Mirror Maker
+
+Deploying an Apache Kafka cluster on Kubernetes using Strimzi is pretty easy.
+For this example, we are going to use `minukube` and just following the corresponding [quickstart](https://strimzi.io/quickstarts/) for deploying the Strimzi operators and provisioning the Apache Kafka cluster.
```
### 
```suggestion
For this example, we are going to use `minukube` and just follow the corresponding [quickstart](https://strimzi.io/quickstarts/) for deploying the Strimzi operators and provisioning the Apache Kafka cluster.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
````diff
@@ -0,0 +1,225 @@
+kubectl apply -f kafka-topic.yaml -n kafka
+```
+
+The Strimzi topic operator takes care of this custom resource in order to create the actual topic on the Apache Kafka cluster.
````
```suggestion
The Strimzi topic operator takes care of this custom resource, creating the topic in the Apache Kafka cluster.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+### Azure Event Hub and the Apache Kafka protocol
+
+Azure Event Hub is a fully managed, real-time data ingestion service provided by Microsoft as part of its Azure cloud offer.
+The artchitectural concepts behind Event Hub are quite the same as Apache Kafka.
```
```suggestion
The artchitectural concepts behind Event Hub are similar to Apache Kafka.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+![Azure Portal: Event Hub creation](/assets/images/posts/2020-05-18-azure-portal-eventhub.png)
+
+Having the Event Hub ready is just the first step.
+In order to configure Kafka Mirror Maker for connecting to the Event Hub, we need to get the corresponding connection string for being authenticated and authorized to write on the Event Hub itself.
```
```suggestion
In order to configure Kafka Mirror Maker for connecting to the Event Hub, we need to get the corresponding connection string in order to be authenticated and authorized to write on the Event Hub itself.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+Having the Event Hub ready is just the first step.
+In order to configure Kafka Mirror Maker for connecting to the Event Hub, we need to get the corresponding connection string for being authenticated and authorized to write on the Event Hub itself.
+Just for simplicity, we can use the already provided `RootManageSharedAccessKey` policy that enable all the rights to manage, send and listen in the Event Hub namespace in which in the `testeh` lives.
```
```suggestion
Just for simplicity, we can use the already provided `RootManageSharedAccessKey` policy that enables all the rights to manage, send and listen in the Event Hub namespace in which in `testeh` lives.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+Having the Event Hub ready is just the first step.
+In order to configure Kafka Mirror Maker for connecting to the Event Hub, we need to get the corresponding connection string for being authenticated and authorized to write on the Event Hub itself.
+Just for simplicity, we can use the already provided `RootManageSharedAccessKey` policy that enable all the rights to manage, send and listen in the Event Hub namespace in which in the `testeh` lives.
+In a real scenario, we would like to create a policy for the mirroring part in order to enable the Kafka Mirror Maker producer just to write to the Event Hub while using a different policy for the application consuming the messages with just the listen right.
```
```suggestion
In a real scenario, we would like to create a policy for the mirroring part in order to enable the Kafka Mirror Maker producer just to write to the Event Hub while using a different policy, with just the permission to listen, for the application consuming the messages. 
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+It means consuming remotely and producing locally.
+In this integration scenario, it is not possible because there is no way to deploy Kafka Mirror Maker alongside the Event Hub namespace that is the target cluster.
+For this reason, the paradigm will be slightly different, deploying Kafka Mirror Maker alongside the source Apache Kafka cluster where we have more control.
+Anyway, it's worth mentioning that the Kubernetes cluster, where Apache Kafka runs, could be an AKS (Azure Kubernetes Service) one so as part of the same Microsoft Cloud infrastracture.
```
```suggestion
Anyway, it's worth mentioning that the Kubernetes cluster, where Apache Kafka runs, could be an AKS (Azure Kubernetes Service) one, part of the same Microsoft Cloud infrastructure at the Event Hub.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+For this reason, the paradigm will be slightly different, deploying Kafka Mirror Maker alongside the source Apache Kafka cluster where we have more control.
+Anyway, it's worth mentioning that the Kubernetes cluster, where Apache Kafka runs, could be an AKS (Azure Kubernetes Service) one so as part of the same Microsoft Cloud infrastracture.
+
+The first step is storing the Event Hub authentication info into a Kubernetes `Secret` in order to be referenced in the Strimzi `KafkaMirrorMaker` custom resource for enabling the producer part to connect to the Event Hub itself.
```
```suggestion
The first step is storing the Event Hub authentication info into a Kubernetes `Secret` in order to be referenced in the Strimzi `KafkaMirrorMaker` custom resource to enable the producer part to connect to the Event Hub itself.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+Anyway, it's worth mentioning that the Kubernetes cluster, where Apache Kafka runs, could be an AKS (Azure Kubernetes Service) one so as part of the same Microsoft Cloud infrastracture.
+
+The first step is storing the Event Hub authentication info into a Kubernetes `Secret` in order to be referenced in the Strimzi `KafkaMirrorMaker` custom resource for enabling the producer part to connect to the Event Hub itself.
+The following snippet shows such a `Secret` where the `$ConnectionString` hasn't to be changed while the endpoint has to be customized with the actual Event Hub connection string; save this `Secret` in a file named `eventhubs-secret.yaml`
```
`$ConnectionString`? I see no `$ConnectionString`. Also "hasn't to be changed" → "must not be changed" or "need not be changed"

ppatierno replied:
I removed a configuration from the snippet but forgot to remove the mention here. Good catch!

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+### Produce, mirror and trigger!
+
+The application consuming the mirrored messages from Event Hub is a simple Azure Functions application which just gets and logs the message that is enough for the purpose of this demo.
```
For the purpose of this demo, the application consuming the mirrored messages from Event Hub is a simple Azure Functions application which just gets and logs the message.

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+Even in this case the `<eventhubs-namespace>` has to be replaced with the Event Hub namespace to connect to and the `EH_CONNECTION_STRING` environment variable has to be defined in the JSON application settings file containing the connection string as already used for the Kafka Mirror Maker configuration.
+
+For more information about Azure Functions connecting to Event Hub this [link](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-hubs-trigger) provides really useful information.
+The application can run even locally or be published on the Azure cloud.
```
```suggestion
The application can run locally or be published on the Azure cloud.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+For more information about Azure Functions connecting to Event Hub this [link](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-hubs-trigger) provides really useful information.
+The application can run even locally or be published on the Azure cloud.
+
+To try the entire pipeline, we just miss sending messages and see how they are mirrored to Event Hub triggering the above function to be executed.
```
just miss sending messages?

"...causing the above function..." would be better than triggering. The "triggering" and the "executed" is a bit tautologous.

ppatierno replied:
I meant that the part we haven't covered yet is sending messages to see how they are mirrored and Azure Functions application is triggered.

ppatierno replied:
> just miss sending messages?

@tombentley what did you mean by that?

sjwoodman replied:
I think @tombentley means that it would read better as "To try the entire pipeline, the only thing left to do is to send some messages. Then we can see that they are mirrored to Event Hub, causing the above function to be executed".

tombentley replied:
@ppatierno sorry I wasn't clear. I meant that I didn't really understand what you were meaning. Simon's suggestion SGTM.

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+The application can run even locally or be published on the Azure cloud.
+
+To try the entire pipeline, we just miss sending messages and see how they are mirrored to Event Hub triggering the above function to be executed.
+To do so, just use the `kafka-console-producer` command linme tool provided by every default Apache Kafka installation.
```
```suggestion
To do so, just use the `kafka-console-producer` command line tool provided with Apache Kafka.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+To try the entire pipeline, we just miss sending messages and see how they are mirrored to Event Hub triggering the above function to be executed.
+To do so, just use the `kafka-console-producer` command linme tool provided by every default Apache Kafka installation.
+Start a new pod in the Kubernetes cluster for hosting the producer and types a couple of JSON messages as follow.
```
```suggestion
Start a new pod in the Kubernetes cluster for hosting the producer and type a couple of JSON messages as follows.
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
````diff
@@ -0,0 +1,225 @@
+>
+```
+
+On the Azure Functions application, the messages will be logged as following.
````
```suggestion
On the Azure Functions application, the messages will be logged like this:
```

tombentley commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+### Conclusion
+
+Integrating applications running on different platforms and different clouds is becoming quite common nowadays and the hybrid cloud based use cases are growing fast.
+Back to the scenario showed in this blog post, the Apache Kafka cluster on Kubernetes could run on any cloud provider (Azure, Amazon, IBM, GCP) and thanks to the Strimzi operators, its deployment and management is fairly simple as well as mirroring data to other systems like Event Hub using Kafka Mirror Maker.
+
+I hope this post has persuaded you of the benefit of using Strimzi and Mirror Maker to do such an integration with Azure.
+Let us know what you are going to integrate!
```
I think it would be interesting to compare this with other ways of doing the same thing (e.g. why are we using MM1 and not MM2? Is there a dedicated Kafka Connector for functions which could provide an alternative?)

ppatierno replied:
Well, MM2 seems to have incompatibility problem at broker protocol level and I would like to dig more into it to figure out, if possible, what they are.
Regarding comparing with other stuff, is it something that should go just in the conclusion section? It seems it could deserve a complete blog post. Or ... what you mean? :-)

ppatierno replied:
@tombentley anything to add here?

tombentley replied:
I think you need to say _something_, because "why did you chose to do it that way" is an obvious question which springs to mind (at least do this reader). It's more obvious given that you're using MM1 and all the momentum is behind MM2, so it seems like a peculiar choice to anyone who's aware of what's going on. If you wanted to commit to blogging about the MM2 alternative then you could say something about the future post, but I can totally understand if you didn't want to commit to that.

sjwoodman replied:
Instead of in the conclusions could you add a note to the EH section (line 72) about the incompatibility? And link to an issue if one exists?

"Note: Whilst MirrorMaker2 is gaining traction in the Kafka ecosystem, there appear to be some protocol compatibility issues with the Azure EventHubs protocol head. Therefore this post uses MirrorMaker1 whilst this is being investigated."

ppatierno replied:
I mentioned that in the EH section as @sjwoodman suggested leaving the conclusion as it is.

sjwoodman commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+The artchitectural concepts behind Event Hub are similar to Apache Kafka.
+An Event Hub is like an Apache Kafka topic and it lives inside an Event Hub namespace that is a kind of Apache Kafka cluster; an Event Hub is partitioned as well.
+Producer applications can publish messages using HTTPS and AMQP 1.0 protocols; consumer applications join consumer groups for receiving messages from the partitions, actually sharing the load across them.
+Few months ago, Azure Event Hub was enriched with an Apache Kafka protocol head (1.0 and above).
```
-> A few months ago

sjwoodman commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+An Event Hub is like an Apache Kafka topic and it lives inside an Event Hub namespace that is a kind of Apache Kafka cluster; an Event Hub is partitioned as well.
+Producer applications can publish messages using HTTPS and AMQP 1.0 protocols; consumer applications join consumer groups for receiving messages from the partitions, actually sharing the load across them.
+Few months ago, Azure Event Hub was enriched with an Apache Kafka protocol head (1.0 and above).
+It enables any Apache Kafka client to connect to an Event Hub, as it was a "normal" Apache Kafka topic, for sending and receiving messages.
```
"as it was" -> "as if it was"

sjwoodman commented on _posts/2020-05-18-mirror-kafka-eventhub.md:
```diff
@@ -0,0 +1,225 @@
+
+### Configure Kafka Mirror Maker
+
+When mirroring messages from a source Apache Kafka cluster to a target one, the usual paradigm is deploying in the Kafka Mirror Maker instance alongside the target cluster.
```
"is deploying in the" -> "is deploying the"

ppatierno commented:
@tombentley @scholzj @sjwoodman I think I addressed all comments but I already had the approval. If there is nothing you want to add, I will publish this blog post tomorrow afternoon. Otherwise, I will fix any more comments you have.
